File: sprint5/By lessons/lesson10code2/lesson10code2.java
import java.util.ArrayList;

public class HeapFunctions {
    public static int popMax(ArrayList<Integer> heap) {
        int result = heap.get(0);
        heap.set(0, heap.get(heap.size() - 1));
        heap.remove(heap.size() - 1);
        siftDown(heap, 0);
        return result;
    }

    public static void siftDown(ArrayList<Integer> heap, int index) {
        int left = 2 * index + 1;
        int right = 2 * index + 2;

        if (left >= heap.size()) {
            return;
        }

        int indexLargest = left;
        if (right < heap.size() && heap.get(left) < heap.get(right)) {
            indexLargest = right;
        }

        if (heap.get(index) < heap.get(indexLargest)) {
            int temp = heap.get(index);
            heap.set(index, heap.get(indexLargest));
            heap.set(indexLargest, temp);
            siftDown(heap, indexLargest);
        }
    }

    public static void main(String[] args) {
        ArrayList<Integer> heap = new ArrayList<>();
        heap.add(8);
        heap.add(5);
        heap.add(3);
        int maxValue = popMax(heap);
        System.out.println(maxValue); // Output: 8
        System.out.println(heap);     // Output: [5, 3]
    }
}

File: sprint5/By lessons/lesson10code2/lesson10code2.go
package main

import "fmt"

func popMax(heap []int) int {
	result := heap[0]
	heap[0] = heap[len(heap)-1]
	heap = heap[:len(heap)-1]
	siftDown(heap, 0)
	return result
}

func siftDown(heap []int, index int) {
	left := 2*index + 1
	right := 2*index + 2

	if len(heap) <= left {
		return
	}

	indexLargest := left
	if right < len(heap) && heap[left] < heap[right] {
		indexLargest = right
	}

	if heap[index] < heap[indexLargest] {
		heap[index], heap[indexLargest] = heap[indexLargest], heap[index]
		siftDown(heap, indexLargest)
	}
}

func main() {
	heap := []int{8, 5, 3}
	maxValue := popMax(heap)
	fmt.Println(maxValue) // Output: 8
	fmt.Println(heap)     // Output: [5 3]
}
